x/evm/handler.go

```diff
@@ -44,82 +46,64 @@ func handleETHTxMsg(ctx sdk.Context, keeper Keeper, msg types.EthereumTxMsg) sdk
-
-	var (
-		leftOverGas uint64
-		addr        common.Address
-		vmerr       error
-		senderRef   = vm.AccountRef(sender)
-	)
-
-	if contractCreation {
-		_, addr, leftOverGas, vmerr = vmenv.Create(senderRef, msg.Data.Payload, usableGas, msg.Data.Amount)
-	} else {
-		// Increment the nonce for the next transaction
-		keeper.csdb.SetNonce(sender, keeper.csdb.GetNonce(sender)+1)
-		_, leftOverGas, vmerr = vmenv.Call(senderRef, *msg.To(), msg.Data.Payload, usableGas, msg.Data.Amount)
-	}
-
-	// handle errors
-	if vmerr != nil {
-		return emint.ErrVMExecution(vmerr.Error()).Result()
+	// Encode transaction by default Tx encoder
+	txEncoder := authutils.GetTxEncoder(types.ModuleCdc)
+	txBytes, err := txEncoder(msg)
+	if err != nil {
+		return sdk.ErrInternal(err.Error()).Result()
 	}
+	txHash := tm.Tx(txBytes).Hash()
 
-	// Refund remaining gas from tx (Check these values and ensure gas is being consumed correctly)
-	refundGas(keeper.csdb, &leftOverGas, msg.Data.GasLimit, context.GasPrice, sender)
+	// Prepare db for logs
+	keeper.csdb.Prepare(common.BytesToHash(txHash), common.Hash{}, *keeper.txCount)
+	*keeper.txCount = *keeper.txCount + 1
 
-	// add balance for the processor of the tx (determine who rewards are being processed to)
-	// TODO: Double check nothing needs to be done here
+	return st.TransitionCSDB(ctx)
+}
 
-	keeper.csdb.Finalise(true) // Change to depend on config
+func handleEmintMsg(ctx sdk.Context, keeper Keeper, msg types.EmintMsg) sdk.Result {
+	if err := msg.ValidateBasic(); err != nil {
+		return err.Result()
+	}
 
-	// TODO: Consume gas from sender
+	// parse the chainID from a string to a base-10 integer
+	intChainID, ok := new(big.Int).SetString(ctx.ChainID(), 10)
+	if !ok {
+		return emint.ErrInvalidChainID(fmt.Sprintf("invalid chainID: %s", ctx.ChainID())).Result()
+	}
 
-	return sdk.Result{Data: addr.Bytes(), GasUsed: msg.Data.GasLimit - leftOverGas}
-}
+	st := types.StateTransition{
+		Sender:       common.BytesToAddress(msg.From.Bytes()),
+		AccountNonce: msg.AccountNonce,
+		Price:        msg.Price.BigInt(),
+		GasLimit:     msg.GasLimit,
+		Amount:       msg.Amount.BigInt(),
+		Payload:      msg.Payload,
+		Csdb:         keeper.csdb,
+		ChainID:      intChainID,
+	}
 
-func refundGas(
-	st vm.StateDB, gasRemaining *uint64, initialGas uint64, gasPrice *big.Int,
-	from common.Address,
-) {
-	// Apply refund counter, capped to half of the used gas.
-	refund := (initialGas - *gasRemaining) / 2
-	if refund > st.GetRefund() {
-		refund = st.GetRefund()
+	if msg.Recipient != nil {
+		to := common.BytesToAddress(msg.Recipient.Bytes())
+		st.Recipient = &to
 	}
-	*gasRemaining += refund
 
-	// Return ETH for remaining gas, exchanged at the original rate.
-	remaining := new(big.Int).Mul(new(big.Int).SetUint64(*gasRemaining), gasPrice)
-	st.AddBalance(from, remaining)
+	// Prepare db for logs
+	keeper.csdb.Prepare(common.Hash{}, common.Hash{}, *keeper.txCount) // Cannot provide tx hash
```

Comment:
I think rather than a counter:

```
i := uint(keeper.csdb.TxIndex())
```

Reply:
That's actually the get method that uses the data set in Prepare. I could add two methods to the csdb interface to allow only tx index to be reset and incremented, but would just change the interface/structure of the csdb from mimmicing geth and is only needed within the Keeper/Cosmos.

Let me know if you want me to bring in 6396abb to make the counter more readable/usable or to add those methods to the csdb type

Reply:
I see. Yeah I would suggest bringing the alternative branch to make it more usable and easier to maintain.

Reply:
awesome! just wanted to check before I did because that commit was a bit subjective